---
title: "Lesniak Lab (Aaron Presser): Peng's siRNA therapy - Seurat script"
output: html_document
---
### I have/will added an R tutorial to this GitHub repository - it will show you how to install and set up R and is relatively comprehensive in showing
### you how to use it, and after you work through it and understand it, you should be competent enough to follow the tutorial here and do a good scRNA-seq ### analysis

### Future references
# Preprocessing single cell data
- This link will teach you a bit more about feature data, how it's used, why, it's used, etc. - it's helpful for understanding what is going on (https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/feature-bc)
- This link will tell you the different functions you have available for preprocessing your data, when you might use each function, and some possible 
  arguments you can set for each function (https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger)
- This link will give you some tutorials on how to use the different preprocessingn frunctions
- There's also an example script in this GitHub repository on how to use the cellranger ```counts``` function - it is similar to the example in the 
  tutorial a couple links up but it may be helpful to see another example
- Here are tutorials on using the Mac/Linux "command line" in your terminal application on your computer, which you will have to use in order to do these 
  preprocessing things (https://www.youtube.com/watch?v=YHFzr-akOas) - for the Windows "command line," look here 
  (https://www.cs.princeton.edu/courses/archive/spr05/cos126/cmd-prompt.html). Also Google around for other pedagogical presentations of the "command 
  line," or if there are things that you don't particular understand how to do (e.g., "linux command line tutorial")


### scRNA-seq analysis
- Basic Seurat tutorial - look here before using this tutorial (https://satijalab.org/seurat/articles/pbmc3k_tutorial.html)
- Helpful for understanding "Integration" (analyzing the control and treatment groups in unison) - (https://satijalab.org/seurat/articles/integration_introduction.html)
- I found this to be a very useful course: (https://github.com/hbctraining/scRNA-seq) - the lessons can be found in the "lessons" folder
- To learn more about data visualization in Seurat, look here (https://satijalab.org/seurat/articles/visualization_vignette.html) and here (https://satijalab.org/seurat/archive/v3.0/visualization_vignette.html)
- Biostars.com is useful for everything bioinformatic (RNA-seq analysis in particular)
- Helpful for understanding the difference between three key differential expression functions, "FindConservedMarkers," "FindMarkers," "FindAllMarkers" - (https://www.biostars.org/p/409790/)
- Overview of basic Seurat workflow, basic Seurat functions, etc. (https://satijalab.org/seurat/articles/essential_commands.html)
- Helpful for understanding "Assays," which will be often used in this tutorial and in practice (https://github.com/satijalab/seurat/wiki/Assay)


### Troubleshooting
- I have/will added an R tutorial to this GitHub repository - check there to see if you are using basic functions correctly
- This is another R tutorial - it is more comprehensive, but will take you longer to go through, and also perhaps to refer back to in the future, but 
  still a very useful resource (here)
- First, with the exception of your dataset having different names, etc. than the ones in this tutorial or on the reference pages, if you are having 
  trouble running a particular line of code, make sure the syntax of a particular line aligns EXACTLY with the syntax used either in this tutorial here, 
  or on the reference page (you can find how to use very Seurat function here: https://satijalab.org/seurat/reference/)
- If you are having an issue running a particular line of code, GOOGLE IT! Someone has likely had your same issue before. If you poke around enough, you   
  will often find an answer on Biostars, StackOverflow, some random message board, or ...
- Seurat's GitHub issue page has a number of questions from people just like you, who are doing things very similar to you. Many of them have been 
  answered - definintely try to find an answer there. If you can't, feel free to ask your question there as well.
- If googling doesn't work, perhaps concurrently with posting on Biostars or Seurat's GitHub issue page or a message board, email the IT department at 
  your institution and ask them for help. They are often incredibly helpful and will solve most of your programming (but not actual analysis!) problems 
  before they will be will be answered elsewhere. Look up "[my institution] research computing" or "[my institution] hpcc office" or "[my institution] it 
  office" etc.
```


# Load data and basic data cleaning - Run this
```{r}
# If there are libraries that are not installed, look at the "Installing packages" part of the basic R tutorial in this repository and look here (https://stackoverflow.com/questions/53336632/what-is-the-code-to-install-biocmanager-package)
### If you want to visualize the locations of lowly expressed genes (like transcription factors, etc.), remove the "#" in front of the line "library(Nebulosa)" and put a "#" in front of the line "library(Seurat)"
#library(Nebulosa)
library(Seurat)
library(ggplot2)
library(dplyr)
library(cowplot)
library(RCurl)
library(AnnotationHub)
library(ensembldb)
library(metap)
library(glmGamPoi)
library(DropletUtils)
library(scran)
library(SingleR)
library(celldex)
library(purrr)
library(tibble)
# library(SeuratDisk)
# library(cowplot)
# library(pagoda2)
# library(loomR)

### If your RNA-seq data was processed by Cellranger, you should have a file on your computer or on your supercomputer (or wherever you data is) called
### "filtered_feature_bc_matrix". If this is your first time using the "terminal" application on Mac/or a linux machine, refer to this video here to
### familiarize yourself with how to access/copy/delete/move/etc. files there: https://www.youtube.com/watch?v=cBokz0LTizk . If you are using a Windows
### machine, refer here: https://www.youtube.com/watch?v=MBBWVgE0ewk
### These next four lines basically import the data from your machine into the analysis software. In the last two lines, with the "min.cells=" argument,
### we filter out genes that are expressed in less than 3 cells. With the "min.features=" argument, we filter out cells in which less than 200 genes are recorded. 
CTRL.data <- Read10X("/projects/p31559/Peng_scRNA_data/cellranger/Control_GEX/outs/filtered_feature_bc_matrix", gene.column=2, unique.features=TRUE)
TREAT.data <- Read10X("/projects/p31559/Peng_scRNA_data/cellranger/Treatment_GEX/outs/filtered_feature_bc_matrix", gene.column=2, unique.features=TRUE)
CTRL <- CreateSeuratObject(counts=CTRL.data, project="CTRL", min.cells=3, min.features=200)
TREAT <- CreateSeuratObject(counts=TREAT.data, project="TREAT", min.cells=3, min.features=200)

### The function "PercentageFeatureSet" calculates the percentage of reads that have some pattern. In this case, we calculate the percentage of CTRL reads 
### that start with the pattern "GRCh38-mt-". The way we tell "PercentageFeatureSet" that only want the reads that start with that pattern is indicated by
### "^", which we put in front of the pattern. If we want only the reads that end with a pattern, we use "$" instead and put it behind the pattern. Since
### each row in the dataframe is an individual cell, we then create a new column in the CTRL dataframe with the percentage of reads matching that pattern
### for each cell (e.g., the percentage of reads beginning with the pattern "MT-" for each cell).
CTRL[["mouse.percent.mt"]] <- PercentageFeatureSet(CTRL, pattern="mt-")
TREAT[["mouse.percent.mt"]] <- PercentageFeatureSet(TREAT, pattern="mt-")

### We can then create violin plots for the values in the columns we created, or any others that are already existing.
VlnPlot(CTRL, features=c("nFeature_RNA", "nCount_RNA", "mouse.percent.mt"), ncol=3, log=TRUE)
VlnPlot(TREAT, features=c("nFeature_RNA", "nCount_RNA", "mouse.percent.mt"), ncol=3, log=TRUE)


### Here, we create new columns where the values are the integer number of gene transcripts matching a pattern in each cell.
CTRL[["MT_genes"]] <- GeneCounter(CTRL, pattern="mt-")
TREAT[["MT_genes"]] <- GeneCounter(TREAT, pattern="mt-")

library(glue)
genes <- list("Actb", "Foxp3", "Odc1", "Cd3d")
for (i in genes) {
  glue(CTRL[["{i}_genes"] <- GeneCounter(CTRL, pattern="i"))
  glue(TREAT[["{i}_genes"] <- GeneCounter(TREAT, pattern="i"))
}

# Delete the following lines if above loop works
#CTRL[["Actb"]] <- GeneCounter(CTRL, pattern="Actb")
#TREAT[["Actb"]] <- GeneCounter(TREAT, pattern="Actb")

#CTRL[["Foxp3"]] <- GeneCounter(CTRL, pattern="Foxp3")
#TREAT[["Foxp3"]] <- GeneCounter(TREAT, pattern="Foxp3")

#CTRL[["Odc1"]] <- GeneCounter(CTRL, pattern="Odc1")
#TREAT[["Odc1"]] <- GeneCounter(TREAT, pattern="Odc1")

#CTRL[["Cd3d"]] <- GeneCounter(CTRL, pattern="Cd3d")
#TREAT[["Cd3d"]] <- GeneCounter(TREAT, pattern="Cd3d")

### We know we will want to filter our cells according to how many or how few unique genes are in each, and also according to how many or how few reads
### are in each cell, so we can create variables that have the values we want to use as our filtering thresholds, and we can plug these variables into the
### following functions where we choose to filter. That way, if we use variables instead of just plugging in the integer values whenever we choose to filter,
### if we choose to change the filtering value, we only need to make the change once (right below, where we define the variables), and that change will
### propagate to each use of the variable. In addition, if we choose to set the filters manually, with the integer value each time, we might make mistakes
### that could affect our analysis significantly and that are very difficult to pinpoint.
min_genes <- 200
max_genes <- 8000
min_reads <- 250
max_reads <- 53000


FeatureScatter(CTRL, feature1="nCount_RNA", feature2="nFeature_RNA") + scale_y_log10() + scale_x_log10() +
  geom_hline(yintercept=min_genes) +
  geom_hline(yintercept=max_genes) +
  geom_vline(xintercept=min_reads) +
  geom_vline(xintercept=max_reads)

CTRL <- subset(x=CTRL, subset=nFeature_RNA > min_genes & nFeature_RNA < max_genes & nCount_RNA > min_reads & nCount_RNA < max_reads & mouse.percent.mt < 9 & mouse.percent.mt > 0)

TREAT <- subset(x=TREAT, subset=nFeature_RNA > min_genes & nFeature_RNA < max_genes & nCount_RNA > min_reads & nCount_RNA < max_reads & mouse.percent.mt < 9 & mouse.percent.mt > 0)
```


# Get Cell Cycle markers - look at CCMs in CTRL and TREAT
```{r}
# These two lines get a list of Mouse (Mus musculus) cell cycle genes separated by cell cycle phase
cc_file <- getURL("https://raw.githubusercontent.com/hbc/tinyatlas/master/cell_cycle/Mus_musculus.csv")
cell_cycle_genes <- read.csv(text=cc_file)

# Connect to AnnotationHub
ah <- AnnotationHub()
# Access to Ensembl database for organism
ahDb <- query(ah,
              pattern = c("Mus Musculus", "EnsDb"),
              ignore.case=TRUE)
# Acquire the latest annotation files
id <- ahDb %>%
  mcols() %>%
  rownames() %>%
  tail(n=1)
# Download the appropriate Ensembld database
edb <- ah[[id]]
# Extract gene-level information from database
annotations <- genes(edb,
                     return.type="data.frame")
# Select annotations of interest
annotations <- annotations %>%
  dplyr::select(gene_id, gene_name, seq_name, gene_biotype, description)
# Get gene names for Ensembl IDs for each gene
cell_cycle_markers <- dplyr::left_join(cell_cycle_genes, annotations, by = c("geneID" = "gene_id"))

# Acquire the S phase genes
s_genes <- cell_cycle_markers %>%
  dplyr::filter(phase=="S") %>%
  pull("gene_name")
# Acquire the G2M phase genes
g2m_genes <- cell_cycle_markers %>%
  dplyr::filter(phase=="G2/M") %>%
  pull("gene_name")
```
```{r}
list_of_experiments <- list(CTRL, TREAT)
### What the following code does is essentially say "for CTRL, run this pipeline, and then for TREAT, run this pipeline
for (i in list_of_experiments) {
  i <- NormalizeData(i)
  i <- CellCycleScoring(i, s.features=s_genes, g2m.features=g2m_genes, set.ident=TRUE)
  i <- FindVariableFeatures(i, selection.method="vst", nfeatures=3000, verbose=FALSE)
  i <- ScaleData(i)
  i <- RunPCA(i)
}
```


# Basic pipeline and integration - Run this - This is a relatively standard pipeline for experiments with CTRL and TREAT groups
```{r}
# The three lines below - for the CTRL dataframe, it creates a column named "groups" and fills it with "CTRLL" all the way down. The TREAT line also
# creates a column named "groups" for the TREAT dataframe, and fills it with "TREAT" all the way down. The third line merges the two dataframes together
CTRL[["groups"]] <- "CTRL"
TREAT[["groups"]] <- "TREAT"
combined <- merge(CTRL, TREAT)

split_seurat <- SplitObject(combined, split.by="groups")
#split_seurat[["human.percent.mt"]] < NULL

# In the few lines below, we split our data by "groups" (i.e., whether any particular piece of data is from the CTRL group or the TREAT group) and then we
# normalize it, and find the 3500 most variable genes
split_seurat <- lapply(X=split_seurat, FUN=function(x) {
  x <- NormalizeData(x)
  x <- FindVariableFeatures(x, method="vst", nfeatures=3500)
})

# In these two lines, we essentially make sure that we are aligning our CTRL data and our TREAT data so when we generate UMAP or PCA plots, the data are
# relatively well aligned. Also, by coordinating the cleaning process for the CTRL and TREAT groups (which is what we are doing here), our analysis will
# generally be more accurate and precise
anchors <- FindIntegrationAnchors(object.list=split_seurat, dims=1:50)
seurat_integrated <- IntegrateData(anchorset=anchors, dims=1:50)

DefaultAssay(seurat_integrated) <- "integrated"

seurat_integrated <- ScaleData(seurat_integrated, verbose=FALSE)
seurat_integrated <- RunPCA(seurat_integrated, npcs=50, verbose=FALSE)

seurat_integrated <- RunUMAP(seurat_integrated, reduction="pca", dims=1:50)
seurat_integrated <- FindNeighbors(seurat_integrated, reduction="pca", dims=1:50)
seurat_integrated <- FindClusters(seurat_integrated, resolution=0.8)

# These lines create (but do not actually plot) plots for the seurat_integrated object - the first one is split into "CTRL" and "TREAT" groups, and the
# second is generated with the clusters labelled
p1 <- DimPlot(seurat_integrated, reduction="umap", group.by="groups")
p2 <- DimPlot(seurat_integrated, reduction="umap", label=TRUE)

# This line actually plots those two different plots
plot_grid(p1, p2)

# Rather than having to run all the above lines of code again to get this processed seurat_integrated object next time we open our computer, we can save
# the seurat_integrated object to our computers and just load it again (fully-processed - see the "readRDS" line below) next time we want to use or
# reference it
saveRDS(seurat_integrated, "/home/atp9753/seurat_integrated.rds")
```
### Should change the file path above and below to your account or p31492

# Read seurat_integrated - If you have already run the pipeline above and you haven't changed seurat_integrated, run this instead of the pipeline
```{r}
seurat_integrated <- readRDS("/home/atp9753/seurat_integrated.rds")
```

# Write seurat_integrated to 10X file - If you want to run SingleR, run this
```{r}
# This line uses the seurat_integrated object to create a new file on your computer than can be input into a piece of software that will automatically
# label the cell types of different cell clustsers - the software is not entirely accurate so you likely shouldn't rely entirely on it
write10xCounts(x=seurat_integrated@assays$RNA@counts, path='seurat_integrated_10X/')
```

```{r
BiocManager::install("scran")
library(scran)
```


# Annotate with SingleR - if you want to run SingleR, run this
```{r}
# Inputting that file created above with the write10xCounts() function into the automatic cell type-labelling software, SingleR
new.data <- Read10X("seurat_integrated_10X")
# Create a reference dataframe so the software has some idea of what different cell types should look like - then if you input unlabelled data into it, it
# can try to predict what cell types those unlabelled data best look like
ref.data <- MouseRNAseqData()

# This makes the actual predictions
predictions <- SingleR(test=new.data, assay.type.test="integrated",
                       ref=ref.data, labels=ref.data$label.main,
                       clusters=seurat_integrated@meta.data$seurat_clusters,
                       
                       de.method="wilcox")
                       
# This presents the predictions in a nice format
table(predictions$labels)
```

```{r}
# If your predictions look reasonable, you can relabel the clusters in your seurat_integrated object to have those predicted names, instead of just
# numbers as they have by default
predicted_labels <- predictions@listData$first.labels
names(predicted_labels) <- levels(seurat_integrated)
seurat_integrated <- RenameIdents(seurat_integrated, predicted_labels)
```

# To visualize the UMAP plot, run this
```{r}
#predicted_labels <- predictions@listData$first.labels
#names(predicted_labels) <- levels(seurat_integrated)
#seurat_integrated <- RenameIdents(seurat_integrated, predicted_labels)
#pdf("seurat_integrated_umap_singler.pdf")
DimPlot(seurat_integrated, reduction="umap", label=TRUE, pt.size=0.5, group.by="groups")
#dev.off()
```

```{r}
# A Nebulosa function (see the code chunk at the top where we loaded different libraries for instructions on running this code chunk) - Nice plots for lowly expressed genes
plot_density(seurat_integrated, features=c("Foxp3", "Odc1"), joint=TRUE) + facet_wrap(.~seurat_integrated$groups)
```


# Cell cluster setting parameters
```{r}
ElbowPlot(seurat_integrated, ndims=100)
```


# To visualize the expression of different genes in the different groups across the cells, run this
```{r}
#pdf("foxp3_il2ra_odc1.pdf")
FeaturePlot(seurat_integrated, features=c("Vegfa", "Itgam", "Cd274"), min.cutoff="q10", split.by="groups")
#dev.off()
```


```{r}
# If there is a population of cells that doesn't fall neatly within one of the k-means clusters but that you want to inspect more, you can manually select them like this
plot <- DimPlot(seurat_integrated, reduction="umap", label=TRUE) + NoLegend()
myeloid_t_cells <- CellSelector(plot)
#write.csv(myeloid_t_cells, "myeloid_t_cells.csv")
```

# Set this up to find all clusters conserved markers
```{r}
# You can then see which genes are conserved between two clusters like this - you can do this for clusters that you manually select with CellSelector() or
# clusters that are generated by way of k-means clustering
#myeloid_t_cells <- read.csv("myeloid_t_cells.csv")
#head(myeloid_t_cells)
Idents(seurat_integrated, cells = myeloid_t_cells) <- "myeloid_t_cells"
DefaultAssay(seurat_integrated) <- "RNA"
all_clusters_conserved_markers <- FindConservedMarkers(seurat_integrated, ident.1=1, grouping.var="groups", verbose=FALSE)
write.csv(mtc_conserved_markers, "all_clusters_conserved_markers.csv")
seurat_integrated@meta.data$seurat_clusters <- seurat_integrated@active.ident
write.csv(seurat_integrated@meta.data, "seurat_integrated_metadata.csv")
```
```{r}
DimPlot(seurat_integrated, reduction="umap", label=T)
```

# myeloid_t_cell markers - vs. clusters 17 and 9
```{r}
# You can then see which genes are differentially expressed in a cluster (relative to another cluster that you specify) like this - you can do this for
# clusters that you manually select with CellSelector() or clusters that are generated by way of k-means clustering
myeloid_t_cell_17_markers <- FindMarkers(seurat_integrated, ident.1="myeloid_t_cells", ident.2="17", min.pct=0.05, logfc.threshold=0.20, test.use="roc")
write.csv(myeloid_t_cell_17_markers, "myeloid_t_cell_17_markers.csv")
```

```{r}
myeloid_t_cell_9_markers <- FindMarkers(seurat_integrated, ident.1="myeloid_t_cells", ident.2="9", min.pct=0.05, logfc.threshold=0.20, test.use="roc")
write.csv(myeloid_t_cell_9_markers, "myeloid_t_cell_9_markers.csv")
```
# Writing seurat_integrated metadata as CSV
```{r}
seurat_integrated@meta.data$seurat_clusters <- seurat_integrated@active.ident
write.csv(seurat_integrated@meta.data, "seurat_integrated_metadata.csv")
```


```{r}
# You can change the identity of individual cell clusters (in your dataframe-data and UMAP) by using the names() function and then the RenameIdents() functions
new.cluster.ids <- c("0", "1", "2", ".", "4", "Tumor", "6", ".", "8", ".", ".", "11", "12", "NK Cells", "14", "15", "16", "CD8 T cells", "cDC1", "Astrocytes/Oligodendrocytes", "20", "21", "cDC2", "23", "24", "25", "26", "Endothelial cells", "28")

#names(new.cluster.ids) <- levels(seurat_integrated)
#seurat_integrated <- RenameIdents(seurat_integrated, new.cluster.ids)
DimPlot(seurat_integrated, reduction="umap", label=TRUE)
```

# '0', '3', '11' sub-populations of macrophages/microglia?
# 'Gm15056' upregulated in macrophages under treatment, 'F10', 'Nos2' downregulated
# 'Ccl3', 'Csf1' cells appear in '9' under treatment
# No memory 'Cd69', 'Itgae' (Cd103) population
# Macrophage4 markers ('Itgam', 'Pf4', 'C1qa', 'C1qb', 'C1qc') seem to shrink under treatment
# Vestigial limb population to the left of '4', '0' disappears under treatment (expresses 'Ly6d', 'Siglech')
# Upper part of '2' possibly astrocytes? - upregulates cell senesncence 'Igfbp4', 'Igfbp7' under treatment 
# '24', left edge of '3' possibly (tumor-associated) endothelial cells? - express 'Col4a1' and 'Igfbp7'
# '21' oligodendrocytes/OPCs - 'Olig2', 'Il33', other markers
# '19' likely cDC2 population - 'Ccr7', 'Ccl5', 'Samsn1', 'Pcgf5', 'Gyg', 'Net1', 'Rapgap1l'
# '18' cDC1-like population - expresses canonical markers 'Itgae', 'Clec9a', but not as much 'Batf3' as e.g. cDC2 population, especially under treatment - surprising?
# 'Il12b', 'Ccl5' appear to be less expressed under treatment, so presumably less migration of cDC1s into TME - less recruitment/activation of Cd8 T cells as a result? - there appears to be fewer 'Cd8a' cells under treatment
# NK cell observations in Google Doc
# Under treatment, '7' and green top part of '14' upregulate 'Il12rb2'
# '8' expresses 'Il1b', 'Ly6c2', 'Sell'

#("Batf3+ Irf8+ DCs")



# Cluster '10' - slight expression of 'Stmn1', 'Mki67' T cell proliferation markers under treatment
# e.g. 'Il1r2' and Il1b' coexpression in treatment and control - 'Il1r2' less in '9', more in '8' in treatment
# Clusters '2', '4,' '6', '14' all seem to have Macrophage4 markers ('Itgam', 'Pf4', 'C1qa', 'C1qb', 'C1qc')
# '13' DC2 - ('Ccr7', 'Ccl5', 'Samsn1', 'Pcgf5', 'Gyg', 'Net1', 'Rabgap1l')
# Right side of '9', 'left bit of '0' potentially Macrophage3 - ('Ly6c2', 'Cxcl9', 'Il1b', H2-Ab1', 'H2-'DMb2', 'Mmp14', 'Cd38')
# Bottom side of '3' potentially Neural cells/NPCs etc. ('Nedd4', 'Neurl3', 'Nenf')



# Run this first
```{r}
### Ignore the details of this function, it basically counts the number of gene transcripts that were recorded for a particular gene
GeneCounter <- function(
  object,
  pattern = NULL,
  features = NULL,
  col.name = NULL,
  assay = NULL
) {
  assay <- assay %||% DefaultAssay(object = object)
  if (!is.null(x = features) && !is.null(x = pattern)) {
    warning("Both pattern and features provided. Pattern is being ignored.")
  }
  features <- features %||% grep(pattern = pattern, x = rownames(x = object[[assay]]), value = TRUE)
  percent.featureset <- colSums(x = GetAssayData(object = object, assay = assay, slot = "counts")[features, , drop = FALSE])
  if (!is.null(x = col.name)) {
    object <- AddMetaData(object = object, metadata = percent.featureset, col.name = col.name)
    return(object)
  }
  return(percent.featureset)
}
```
